cpu/sam0_common/periph/flashpage.c

```diff
@@ -60,189 +60,205 @@ static void _unlock(void)
-#ifdef FLASHPAGE_RWWEE_NUMOF
-    if (flash_type == FLASH_RWWEE) {
-#ifdef CPU_SAML1X
-         /* SAML1X use the same Write Page command for both flash memories */
-        _NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_WP);
-#else
-        _NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_RWWEEWP);
-#endif
-    } else {
-#endif
+
+    /* send Row/Block erase command */
 #ifdef NVMCTRL_CTRLB_CMDEX_KEY
-        _NVMCTRL->CTRLB.reg = (NVMCTRL_CTRLB_CMDEX_KEY | NVMCTRL_CTRLB_CMD_WP);
+    _NVMCTRL->CTRLB.reg = (NVMCTRL_CTRLB_CMDEX_KEY | NVMCTRL_CTRLB_CMD_EB);
 #else
-        _NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_WP);
+    _NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_ER);
 #endif
-#ifdef FLASHPAGE_RWWEE_NUMOF
-    }
-#endif
-    wait_nvm_is_ready();
-    _lock();
 }
 
-#ifdef FLASHPAGE_RWWEE_NUMOF
-void flashpage_write_internal(int page, const void *data, int flash_type)
+static void _cmd_write_page(void)
+{
+    wait_nvm_is_ready();
+
+    /* write page */
+#ifdef NVMCTRL_CTRLB_CMDEX_KEY
+    _NVMCTRL->CTRLB.reg = (NVMCTRL_CTRLB_CMDEX_KEY | NVMCTRL_CTRLB_CMD_WP);
 #else
-void flashpage_write(int page, const void *data)
+    _NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_WP);
 #endif
+}
+
+static void _write_page(void* dst, const void *data, size_t len, void (*cmd_write)(void))
 {
-    uint32_t *page_addr;
+    uint32_t *dst32 = dst;
 
-#ifdef FLASHPAGE_RWWEE_NUMOF
-    if (flash_type == FLASH_RWWEE) {
-        page_addr = (uint32_t *)flashpage_rwwee_addr(page);
-    } else {
-#endif
-        page_addr = (uint32_t *)flashpage_addr(page);
-#ifdef FLASHPAGE_RWWEE_NUMOF
+    _unlock();
+    _cmd_clear_page_buffer();
+
+    /* copy whole words */
+    const uint32_t *data32 = data;
+    while (len) {
+        *dst32++ = *data32++;
+        len -= sizeof(uint32_t);
     }
-#endif
+
+    cmd_write();
+    _lock();
+}
+
+static void _erase_page(void* page, void (*cmd_erase)(void))
+{
+    uintptr_t page_addr = (uintptr_t)page;
 
     /* erase given page (the ADDR register uses 16-bit addresses) */
     _unlock();
-#if defined(CPU_SAML1X) || defined(CPU_SAMD5X)
-    /* Ensure address alignment */
-    _NVMCTRL->ADDR.reg = (((uint32_t)page_addr) & 0xfffffffe);
-#else
-    _NVMCTRL->ADDR.reg = (((uint32_t)page_addr) >> 1);
-#endif
-#ifdef FLASHPAGE_RWWEE_NUMOF
-    if (flash_type == FLASH_RWWEE) {
-#ifdef CPU_SAML1X
-         /* SAML1X use the same Erase command for both flash memories */
-        _NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_ER);
-#else
-        _NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_RWWEEER);
-#endif
-    } else {
-#endif
-#ifdef NVMCTRL_CTRLB_CMDEX_KEY
-        _NVMCTRL->CTRLB.reg = (NVMCTRL_CTRLB_CMDEX_KEY | NVMCTRL_CTRLB_CMD_EB);
-#else
-        _NVMCTRL->CTRLA.reg = (NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_ER);
-#endif
-#ifdef FLASHPAGE_RWWEE_NUMOF
-    }
+
+    /* ??? */
```

Comment:
There is this comment in the reference manual: `ADDR drives the hardware (16-bit) address to the NVM when a command is executed using CMDEX. 8-bit
addresses must be shifted one bit to the right before writing to this register.` But are we using 8bit addresses here?

Reply:
`Write the address of the row to erase ADDR. Any address within the row can be used., I don't quite understand this sift.

Reply:
Ah that makes sense.
I would interpret it as such that internally the device addresses half-words (16-bit), but we are addressing bytes (8-bit), so we must shift by one.

Reply:
can you add a comment?

Reply:
added

Reply:
Addressed